bread/layout/components/search_select.py

```diff
@@ -0,0 +1,109 @@
+import htmlgenerator as hg
+from django.utils.translation import gettext_lazy as _
+
+import bread.utils
+
+from .icon import Icon
+from .loading import Loading
+from .tag import Tag
+
+
+class SearchSelect(hg.BaseElement):
+    def __init__(
+        self,
+        search_view,
+        query_urlparameter="q",
+        size="xl",
+        searchinput_widgetattributes=None,
+        widgetattributes=None,
+        **elementattributes,
+    ):
+        current_selection_id = widgetattributes["value"][0]
+        # This works inside a formset. Might need to be changed for other usages.
+        current_selection = (
+            getattr(
+                elementattributes["boundfield"].form.instance,
+                elementattributes["fieldname"],
+            )
+            if current_selection_id
+            else None
+        )
+        del elementattributes["boundfield"]
+        elementattributes["_class"] = (
+            elementattributes.get("_class", "") + f" bx--search bx--search--{size}"
+        )
+        elementattributes["data_search"] = True
+        elementattributes["role"] = "search"
+
+        search_widget_attributes = {
+            "id": "search__" + hg.html_id(self),
+            "_class": "bx--search-input",
+            "type": "text",
+            "style": "width: auto;",
+            **(searchinput_widgetattributes or {}),
+        }
+
+        resultcontainerid = f"search-result-{hg.html_id((self, search_view))}"
+
+        url = bread.utils.reverse(
+            search_view,
+            query={
+                "target-id-to-store-selected": widgetattributes["id"],
+            },
+        )
```

Comment:
As mentioned above: Possible to move into basxconnect?

Reply:
The problem is that this ID needs to be unique for each `SearchSelect` widget that is created. But the widgets themselves are not created in basxconnect, but in `_mapwidget`...

Reply:
Okay, I would like us to work a bit on this one.

Apart from the viewname vs URL. issue we also have the problem that the name of the query parameter needs to be defined/hardcoded in two different places, once in ```layout/components/search_select.py``` and once in ```basxconnect/core/views/person/person_search_views.py```.

Try to decouple the search component from the search view a bit more would improve things I suggest. How about something like this:

- The search view can return whatever it wants to be. We just need to make sure that the pk-values and the labels are extractable via CSS selectors.
- The search component takes as ```__init__``` parameter a URL, and three CSS-selector strings:
  - item_selector: selects an item (which will get a click-handler attached)
  - item_label_selector: selects a label inside a an item
  - item_value_selector: selects the value inside an item (simply the value which goes into the hidden field, not necessarily a pk)
  I think with that we can basically cover all scenarios and also use any source (which allows CORS requests) with this.
- The search components uses htmx.onLoad (https://htmx.org/docs/#events) when loaded (using ```onload="htmx.onLoad(....)"```) to add a listener for new search results coming in.
- The listener attaches onclick handlers to all items which match item_selector (maybe using ```$$({item_selector})._.bind(...)```?). The onclick handlers should update the label and the value of the component by using item_label_selector and item_value_selector to extract the correct values. Getting e.g. the value would be something like ```f"$({item_value_selector}, this)"``` assuming this is inside a python string

I did not carve it out completely, it is just a draft of how I would remove the coupling between search element and search view.
What do you think? If you think it is too much effort for now I could work on this as well.

Reply:
I think this is a great suggestion!

Reply:
I am also thinking whether it would not be nicer to have the `person_search_views` just return JSON and do the rest inside the component.

Reply:
Yes, JSON might be appropriate for this use case. I am not sure if we would want to make a full swing to JSON only though (it requires some decisions to be made and converting JSON to HTML is not always as straight-forward as we would hope it to be), but adding a JSON option will likely become necessary if we grow the project...
The reason I avoided it so far is because if we introduce "client-side templating" (meaning getting JSON results and rendering them into some html on the client side) we need to think about how we implement that well. For a single case some quick-n-dirt javascript to generate html would be okay. But it can become a bit tedious if we take the concept further. And at that point we would want to use a javascript-templating library like mustache or handlebars. And so it can lead down a bit of a rabbit-trail :)
In any case, I woul add the css-selector concept (so that it is possible to use views which were not created with autocomplete in mind) and keep the option (via arguments or as a second autocomplete component) to work with JSON responses.